### PR TITLE
Store the demographics survey data and send the related ping

### LINKED ISSUE #124: implement add-on storage + submission for demographic survey
TO finish the onboarding flow implementation, we'll need to submit the demographic survey data provided by the user.

1. The current schema is embedded [in the demographic page survey](https://github.com/mozilla-ion/ion-core-addon/blob/master/src/routes/demographics/Content.svelte#L57). I am guessing it will need to be removed from here and made into a separate json file so we can use it elsewhere.
2. we'll need to add store functions in `create-store.js`, `web-extension.js`, and the background scripts.

### PR DESCRIPTION
**Pull request by Dexterp37**
Fixes #124

It stores the survey data in the core addon and defines a ping to send out.

**Important**: the schema will be deployed in #182.

Todo:

- [x] Write unit tests.
- [x] Design the survey ping and send it.
- [x] Get encryption keys for the pioneer-meta table.
- [x] Document the new ping.
- [x] ~~Deploy the new ping schema.~~

### CODE REVIEW

File: src/routes/demographics/Content.svelte [hunk 1]
```diff
@@ -65,12 +65,12 @@
       type: "single",
       columns: true,
       values: [
-        { key: "19-24", label: "19-24 years old" },
-        { key: "25-34", label: "25-34 years old" },
-        { key: "35-44", label: "35-44 years old" },
-        { key: "45-54", label: "45-54 years old" },
-        { key: "55-64", label: "55-64 years old" },
-        { key: "> 65", label: "65 years and older" },
+        { key: "19_24", label: "19-24 years old" },
```
Comment (Dexterp37) [inline src/routes/demographics/Content.svelte:68]: @hamilton this PR changes the keys here, to make it easier to work with them data collection-wise.

File: docs/DATA_COLLECTION.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# Data Documentation
```
Comment (Dexterp37) [inline docs/DATA_COLLECTION.md:1]: @hamilton @knowtheory FYI, here's the data documentation for the core addon.

**Review by hamilton — COMMENTED** [review]
Looked back on @Dexterp37 's comments toward me – the Svelte component changes look good, so marking my review as "comment". I think @rhelmer can probably do a more thorough review of the code.

File: core-addon/DataCollection.js [hunk 3]
```diff
@@ -156,4 +156,68 @@ module.exports = class DataCollection {
...
+    // possible values).
+    for (let originalField in FIELD_MAPPING) {
+      if (originalField in data) {
+        const newName = FIELD_MAPPING[originalField];
+        if (!(newName in processed)) {
+          processed[newName] = {};
+        }
+        processed[newName][data[originalField]] = true;
+      }
+    }
```
Comment (rhelmer) [inline core-addon/DataCollection.js:202]: Hm, each key in `data` will be unique so I'm not sure why it needs to check if the key already exists in processed?

Would something likethis work?

```suggestion
for (const [originalField, newName] of Object.entries(FIELD_MAPPING)) {
  if (originalField in data) {
    processed[newName] = { [data[originalField]]: true };
  }
}
```

File: core-addon/DataCollection.js [hunk 4]
```diff
@@ -156,4 +156,68 @@ module.exports = class DataCollection {
...
+    // Note: "race" gets renamed to "races" and has multiple
+    // values.
+    if ("race" in data) {
+      for (let race of data["race"]) {
+        if (!("races" in processed)) {
+          processed["races"] = {};
+        }
+        processed["races"][race] = true;
+      }
+    }
```
Comment (rhelmer) [inline core-addon/DataCollection.js:213]: Could this be:
```suggestion
    if ("race" in data) {
      processed["races"] = data.race.reduce((a, b) => ((a[b] = true), a), {});
    }
```

**Review by rhelmer — APPROVED** [review]
A few suggestions but feel free to take or leave, lgtm!

File: core-addon/DataCollection.js (hunk 3 above)
Comment (Dexterp37) [reply core-addon/DataCollection.js:202]: Oh! Good point!

**Comment by Dexterp37** [comment]
Flagging @tdsmith for data-review.

For some additional context: this is sending a ping for collecting _optional_ demographics information (i.e. as outlined [in this schema](https://github.com/mozilla-ion/ion-core-addon/blob/da0cebb1ccd97d3a806718f596de86f748422df6/src/routes/demographics/Content.svelte#L61-L176)) from users who opt into the Rally platform.

# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

- What's the composition of the group of users joining Rally?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:

The Mozilla Rally platform needs to answer the question above in order to scientifically frame the results of the studies that are running on the Rally platform.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

We have no other platform or way to collect this kind of information.

4) Can current instrumentation answer these questions?

No.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

The list of collected metrics is available in the `DATA_DOCUMENTATION.md` file. Here's a [convenient link](https://github.com/mozilla-ion/ion-core-addon/pull/139/files#diff-5b979597e2289343bc99a6c10c9db394e93eb80e6013d31ec325c132c32bf6f8R43) to the relevant section.

The above is probably either Category 3 or Category 4. I'm not sure which one, though.

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

Documentation for these metrics is part of this PR (`DATA_DOCUMENTATION.md` and will permanently live in the `docs/` directory of this repository).

7) How long will this data be collected?  Choose one of the following:

Permanently. @knowtheory and @rjweiss are responsible for this data.

8) What populations will you measure?

The population of users joining the Rally platform.

9) If this data collection is default on, what is the opt-out mechanism for users?

The data is not default on.
After installing this addon, user is presented with a page to join the platform.
If the user joins, then they are presented with an optional demographics questionnaire that can be skipped.
No data is collected if the questionnaire is skipped.

10) Please provide a general description of how you will analyze this data.

The data will be used to statistically describe the panel of users joining the Ion/Rally platform.

11) Where do you intend to share the results of your analysis?

Aggregate reports, anonymous statistics created from the collected data may be published with research papers from our partners, on the Rally website or internally.

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:

No. Data will be analysed in the Mozilla Rally Analysis environment and collected through Legacy Telemetry (and Glean in the future).

**Comment by tdsmith** [comment]
This is a higher-category collection so I'll escalate to our sponsor, @litkaemily.

@litkaemily, can you confirm that this collection is permissible for Rally?

**Comment by litkaemily** [comment]
@knowtheory I wan to confirm, are you using any new vendors to support the survey (i.e., will data be stored in a new place or in a new system)? Also, we'll need to update the rally privacy notice + sumo articles; when is this to go live?

**Comment by knowtheory** [comment]
Hey @litkaemily,

I _think_ that's out of scope for this particular issue (just for everyone else's sake), but just quickly to recap my understanding of our state:

We have a demographic profile that we're asking users to fill out.  That's what we're calling our initial user survey (and is the scope for the current issue), all the data there is (potentially) collected through the Core Addon enrollment experience, and goes into our platform directly (through code we've written ourselves, and through our pipes).

Separately, some of our partners may wish to ask questions outside the scope of our core profile data collection, and for that, we're providing a mechanism where they can prompt users to fill out _separate_ per-study surveys, which may be implemented through 3rd party platforms (like Qualtrics).  Any such surveys will be flagged individually in the study review & approval process.

So, tl;dr: Rally program data isn't going into 3rd party tools; individual studies may ask for 3rd party survey tools.

**Comment by litkaemily** [comment]
Thanks for explaining! I'm fine with the data collection, so long as we update the privacy policy/other disclosures prior to shipping.

**Comment by tdsmith** [comment]
In that case, r+:

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, as provided.

2) Is there a control mechanism that allows the user to turn the data collection on and off?

n/a, opt-in collection.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

@knowtheory and @rjweiss are responsible for this data.

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Demographic data doesn't neatly fit any of our categories since it's not amenable to mechanical collection but we'll round up to "highly sensitive," Cat 4.

5) Is the data collection request for default-on or default-off?

It's an opt-in collection.

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No.

7) Is the data collection covered by the existing Firefox privacy notice?

The Pioneer privacy notice should be updated before shipping, per @litkaemily.

8) Does there need to be a check-in in the future to determine whether to renew the data?

n/a

9) Does the data collection use a third-party collection tool?

No.

**Comment by Dexterp37** [comment]
> 
> 
> Thanks for explaining! I'm fine with the data collection, so long as we update the privacy policy/other disclosures prior to shipping.

Thank you @knowtheory @litkaemily @tdsmith . I filed mozilla-ion/internal#10 to track this (cc @marniepw ) as a blocker for shipping.
